### ISSUE TITLE
Template parsing performance degradation between 0.8.3.3 and 0.9.1.0

### ISSUE THREAD

**Issue by ugenl** [bkiers/Liqp#310]
My team has observed notably slower (between 5x and 8x) parsing times when parsing large template strings in v0.9.1.0 as compared to v0.8.3.3. 

The example template string I've attached is ~3.6k lines long, and largely consists of massive `if`/`else` blocks and `assign`s. We ran a test that parsed this string 500 times and averaged the time taken, with the following results:

0.8.3.3 - 0.0402 seconds
0.9.1.0 - 0.2390 seconds

Here's the kotlin test that we were using:

```
fun massiveTemplateParse() {
        val templateString = File(<filepath>).readText()
        val templateParser = TemplateParser.Builder().build()
        val timeArray = mutableListOf<Double>()
        var template: Template?

        repeat(500) {
            val startTime = System.currentTimeMillis()
            try {
                template = templateParser.parse(templateString)
            } catch (e: LiquidException) {
                val errorMessage =
                    LiquidTemplate.cleanParseExceptionForLookml(e.message, templateString)
                throw LiquidParseException("Liquid parse exception: $errorMessage", e)
            }
            val timeTaken = System.currentTimeMillis() - startTime
            Assert.assertNotNull(template)
            timeArray.add(timeTaken/1000.0)
            println("Time taken: ${timeTaken/1000.0}")
        }
        timeArray.removeFirst()
        timeArray.removeFirst()
        timeArray.removeFirst()
        println("Avg time taken: ${timeArray.average()}")
    }
```

Also attached is an archive with yourkit profiling snapshots of each test run. It seems like there are simply more function invocations for 0.9.1.0, but I wasn't able to dig much further than that.

[test_template_string.txt](https://github.com/user-attachments/files/17264247/test_template_string.txt)
[snapshots.zip](https://github.com/user-attachments/files/17264248/snapshots.zip)

**Comment by vtatai** [comment]
I did a git bisect and the problematic commit is `bec3a250d87a81da044a25102b485dfaef4b9334`.

**Comment by bkiers** [comment]
> I looked the bug but my knowlage is not enought to see exact cause. All I found it is most probably related with this change: ![image](https://private-user-images.githubusercontent.com/773163/375657814-de204d18-3936-4a42-a379-875b24e9b67e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MzE3MjYsIm5iZiI6MTcyODYzMTQyNiwicGF0aCI6Ii83NzMxNjMvMzc1NjU3ODE0LWRlMjA0ZDE4LTM5MzYtNGE0Mi1hMzc5LTg3NWIyNGU5YjY3ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMVQwNzIzNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZjBiYjNlZmQ4OTQ5MmEzMTYwMzYyMWU2NGZlNTdlMWM2NjcyNDQwOWRkYmRmMWVhZDNiM2YyMjc3OWIxMTg0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.dTmwbVd5qSeC4OmB3BupaqsxQRPxIWqlj-i7oq3ImSY) but here need assistance from @bkiers

Hmm, seems odd that these changes would cause such a decline in performance. Will have a closer look this weekend.

**Comment by bkiers** [comment]
Yes, I see that the rule:

```
output
 : {isEvaluateInOutputTag()}? outStart evaluate=expr filter* OutEnd
 | {isStrict()}? outStart term filter* OutEnd
 | {isWarn() || isLax()}? outStart term filter* unparsed=not_out_end? OutEnd
 ;

not_out_end
 : ( ~OutEnd )+
 ;
```

is the culprit, which is odd given that `isStrict() == true`. When commenting out `| {isWarn() || isLax()}? outStart term filter* unparsed=not_out_end? OutEnd` (and also fixing the error in NodeVisitor), the performance is much better again. Will need some more investigating as to why... 🤔

**Comment by bkiers** [comment]
Look like I have a fix @msangel  and @ugenl

**Comment by msangel** [comment]
Published to maven as `0.9.1.1` version. Soon it will be available in maven central.
